less/scripts/get_info/grad/get_train_lora_grads.py

```diff
@@ -0,0 +1,43 @@
+import os
+import argparse
+import subprocess
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--train_data_name", type=str, required=True, help="Name of the training data (will be used to store gradients)")
+    parser.add_argument("--train_file", type=str, required=True, help="Path to training file")
+    parser.add_argument("--model_path", type=str, required=True, help="Path to model in the `out` directory, e.g. 'llama2-7b-p0.05-lora-seed3'")
+    parser.add_argument("--ckpts", type=str, required=True, help="List of checkpoints to compute gradients for, e.g. '105 211 317 420'")
+    parser.add_argument("--dims", default=8192, type=int, required=False, help="Dimension of projection")
+    args = parser.parse_args()
+
+    # Split checkpoint string into list of ints
+    ckpts = [int(x) for x in args.ckpts.split()]
+
+    # Process each checkpoint
+    for ckpt in ckpts:
+        # Construct model path with checkpoint
+        model_checkpoint_path = os.path.join("../out", args.model_path, f"checkpoint-{ckpt}")
+        
+        # Construct output path with checkpoint
+        output_path = os.path.join("../grads", args.model_path, f"{args.train_data_name}-ckpt{ckpt}-adam")
+
+        # Create output directory if it doesn't exist
+        os.makedirs(output_path, exist_ok=True)
+
+        # Build command
+        cmd = [
+            "python3", "-m", "less.data_selection.get_info",
+            "--train_file", args.train_file,
+            "--info_type", "grads",
+            "--model_path", model_checkpoint_path,
+            "--output_path", output_path,
+            "--gradient_projection_dimension", str(args.dims),
+            "--gradient_type", "adam"
```

Comment:
Sanity check question: Adam and AdamW provide the same type of gradient data, am I right?

Reply:
In this script in particular AdamW is used across all computations (despite this variable being just "Adam") -- it's used for warming up the model and for getting the gradients for the train data. For obtaining the gradients of eval data, gradient_type is set to "sgd" 


Here is the difference in computation: 
- SGD 
```
def obtain_gradients(model, batch):
    """ obtain gradients. """
    loss = model(**batch).loss
    loss.backward()
    vectorized_grads = torch.cat(
        [p.grad.view(-1) for p in model.parameters() if p.grad is not None])
    return vectorized_grads
``` 

- Adam(W) 
```
def prepare_optimizer_state(model, optimizer_state, device):
    names = [n for n, p in model.named_parameters() if p.requires_grad]
    avg = torch.cat([optimizer_state[n]["exp_avg"].view(-1) for n in names])
    avg_sq = torch.cat([optimizer_state[n]["exp_avg_sq"].view(-1)
                       for n in names])
    avg = avg.to(device)
    avg_sq = avg_sq.to(device)
    return avg, avg_sq


def obtain_gradients_with_adam(model, batch, avg, avg_sq):
    """ obtain gradients with adam optimizer states. """
    beta1 = 0.9
    beta2 = 0.999
    eps = 1e-08

    loss = model(**batch).loss
    loss.backward()

    vectorized_grads = torch.cat(
        [p.grad.view(-1) for n, p in model.named_parameters() if p.grad is not None])

    updated_avg = beta1 * avg + (1 - beta1) * vectorized_grads
    updated_avg_sq = beta2 * avg_sq + (1 - beta2) * vectorized_grads ** 2
    vectorized_grads = updated_avg / torch.sqrt(updated_avg_sq + eps)

    return vectorized_grads
 ```
 
 Looking at this, I think in principle this code would work with a normal Adam, but the gradient values produces by Adam and AdamW would be different